sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java

```diff
@@ -450,53 +519,136 @@ static RunQueryRequest makeRequest(Query query, @Nullable String namespace) {
               .withLabel("Namespace"))
-          .addIfNotNull(DisplayData.item("query", getQuery().toString())
-              .withLabel("Query"));
+          .addIfNotNull(DisplayData.item("query", query)
+              .withLabel("Query"))
+          .addIfNotNull(DisplayData.item("gqlQuery", getGqlQuery()));
     }
 
-    /**
-     * A class for v1 Cloud Datastore related options.
-     */
     @VisibleForTesting
     static class V1Options implements Serializable {
-      private final Query query;
-      private final String projectId;
+      private final ValueProvider<String> project;
       @Nullable
-      private final String namespace;
+      private final ValueProvider<String> namespace;
 
-      private V1Options(String projectId, Query query, @Nullable String namespace) {
-        this.projectId = checkNotNull(projectId, "projectId");
-        this.query = checkNotNull(query, "query");
+      private V1Options(ValueProvider<String> project, ValueProvider<String> namespace) {
+        this.project = project;
         this.namespace = namespace;
       }
 
-      public static V1Options from(String projectId, Query query, @Nullable String namespace) {
-        return new V1Options(projectId, query, namespace);
+      public static V1Options from(String projectId, String namespace) {
+        return from(StaticValueProvider.of(projectId), StaticValueProvider.of(namespace));
       }
 
-      public Query getQuery() {
-        return query;
+      public static V1Options from(ValueProvider<String> project, ValueProvider<String> namespace) {
+        return new V1Options(project, namespace);
       }
 
       public String getProjectId() {
-        return projectId;
+        return project.get();
       }
 
       @Nullable
       public String getNamespace() {
+        return namespace == null ? null : namespace.get();
+      }
+
+      public ValueProvider<String> getProjectValueProvider() {
+        return project;
+      }
+
+      @Nullable
+      public ValueProvider<String> getNamespaceValueProvider() {
         return namespace;
       }
+
+    }
+
+    /**
+     * A DoFn that translates a Cloud Datastore gql query string to {@code Query}.
+     */
+    static class GqlQueryTranslatorFn
+        extends DoFn<ValueProvider<String>, Query> {
+      private final V1Options options;
+      private transient Datastore datastore;
+      private final V1DatastoreFactory datastoreFactory;
+
+      GqlQueryTranslatorFn(V1Options options) {
+        this(options, new V1DatastoreFactory());
+      }
+
+      @VisibleForTesting
+      GqlQueryTranslatorFn(V1Options options, V1DatastoreFactory datastoreFactory) {
+        this.options = options;
+        this.datastoreFactory = datastoreFactory;
+      }
+
+      @StartBundle
+      public void startBundle(Context c) throws Exception {
+        datastore = datastoreFactory.getDatastore(c.getPipelineOptions(), options.getProjectId());
```

Comment:
This is just for future proof in case we need to run it for multiple queries, would like to keep it given that is not too much code compared to creating a `SimpleFunction`.